packages/type-utils/tests/TypeOrValueSpecifier.test.ts

```diff
@@ -0,0 +1,306 @@
+      [
+        {
+          from: 'package',
+          name: ['MyType', 'MyType'],
+          source: './filename.js',
+        },
+      ],
+      [{ from: 'package', name: [], source: './filename.js' }],
+      [{ from: 'package', name: 'MyType' }],
+      [{ from: 'package', source: './filename.js' }],
+      [{ from: 'package', name: 'MyType', source: 42 }],
+      [
+        {
+          from: 'package',
+          name: ['MyType', 'myValue'],
+          source: ['./filename.js', './another-file.js'],
+        },
+      ],
+      [
+        {
+          from: 'package',
+          name: 'MyType',
+          source: './filename.js',
+          unrelatedProperty: '',
+        },
+      ],
+    ])("doesn't match a malformed package specifier", runTestNegative);
+
+    it.each([
+      [{ from: ['file'], name: 'MyType' }],
+      [{ from: ['lib'], name: 'MyType' }],
+      [{ from: ['package'], name: 'MyType' }],
+      [{ from: ['file', 'lib'], name: 'MyType' }],
+      [{ from: ['file', 'package'], name: 'MyType' }],
+      [{ from: ['lib', 'package'], name: 'MyType' }],
+      [{ from: ['file', 'lib', 'package'], name: 'MyType' }],
+    ])('matches a multi-source specifier', runTestPositive);
+
+    it.each([
+      [{ from: [], name: 'MyType' }],
+      [{ from: ['invalid'], name: 'MyType' }],
+      [{ from: ['file', 'invalid'], name: 'MyType' }],
+      [{ from: ['file', 'file'], name: 'MyType' }],
+      [{ from: ['file'], name: 42 }],
+      [{ from: ['file'] }],
+      [{ from: ['file'], name: 'MyType', unrelatedProperty: '' }],
+    ])("doesn't match a malformed multi-source specifier", runTestNegative);
+  });
+
+  describe('typeMatchesSpecifier', () => {
+    function runTests(
+      code: string,
+      specifier: TypeOrValueSpecifier,
+      expected: boolean,
+    ): void {
+      const rootDir = path.join(__dirname, 'fixtures');
+      const { ast, services } = parseForESLint(code, {
+        project: './tsconfig.json',
+        filePath: path.join(rootDir, 'file.ts'),
+        tsconfigRootDir: rootDir,
+      });
+      const type = services.program
+        .getTypeChecker()
+        .getTypeAtLocation(
+          services.esTreeNodeToTSNodeMap.get(
+            (ast.body[0] as TSESTree.TSTypeAliasDeclaration).id,
+          ),
+        );
+      expect(typeMatchesSpecifier(type, specifier, services.program)).toBe(
+        expected,
+      );
+    }
+
+    function runTestPositive(
+      code: string,
+      specifier: TypeOrValueSpecifier,
+    ): void {
+      runTests(code, specifier, true);
+    }
+
+    function runTestNegative(
+      code: string,
+      specifier: TypeOrValueSpecifier,
+    ): void {
+      runTests(code, specifier, false);
+    }
+
+    it.each<[string, TypeOrValueSpecifier]>([
+      ['interface Foo {prop: string}; type Test = Foo;', 'Foo'],
+      ['type Test = RegExp;', 'RegExp'],
+    ])('correctly matches a universal string specifier', runTestPositive);
```

Comment:
[Nit] _correctly_ is implied by it being a test. 🙂 Here and elsewhere.

```suggestion
    ])('matches a matching universal string specifier', runTestPositive);
```